### PR TITLE
Add log Spark function

### PR DESCRIPTION
**Pull request by zhli1142015**
Doc:https://spark.apache.org/docs/latest/api/sql/#log
Code:https://github.com/apache/spark/blob/f0b7cfa56cb90ef70132d9656299956cbde00b53/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala#L1401

### CODE REVIEW

**Comment by zhli1142015** [comment]
cc @rui-mo  and @PHILO-HE , thanks.

File: velox/docs/functions/spark/math.rst [hunk 1]
```diff
@@ -147,6 +147,11 @@ Mathematical Functions
     Returns true if x is Nan, or false otherwise. Returns false is x is NULL.
     Supported types are: REAL, DOUBLE.
 
+.. spark:function:: log(base, expr) -> double
+
+    Returns the logarithm of ``expr`` with ``base``..
```
Comment (rui-mo) [inline velox/docs/functions/spark/math.rst:152]: nit: remove duplicate `.`

File: velox/docs/functions/spark/math.rst [hunk 2]
```diff
@@ -147,6 +147,11 @@ Mathematical Functions
...
+    Return NULL if either ``expr`` or ``base`` is less than or equal to 0.
```
Comment (rui-mo) [inline velox/docs/functions/spark/math.rst:153]: nit: Return -> Returns

File: velox/functions/sparksql/Arithmetic.h [hunk 3]
```diff
@@ -262,6 +262,17 @@ struct Log1pFunction {
   }
 };
 
+template <typename T>
+struct LogarithmFunction {
+  FOLLY_ALWAYS_INLINE bool call(double& result, double a, double b) {
+    if (a <= 0 || b <= 0) {
+      return false;
+    }
+    result = std::log(b) / std::log(a);
```
Comment (rui-mo) [inline velox/functions/sparksql/Arithmetic.h:271]: Can we add a test for `std::log(a) == 0`?

**Review by rui-mo — COMMENTED** [review]
Thanks.

File: velox/functions/sparksql/Arithmetic.h (hunk 3 above)
Comment (zhli1142015) [reply velox/functions/sparksql/Arithmetic.h:271]: Added, thanks

**Review by PHILO-HE — APPROVED** [review]
Looks good to me. Thanks!

File: velox/functions/sparksql/Arithmetic.h [hunk 4]
```diff
@@ -262,6 +262,17 @@ struct Log1pFunction {
   }
 };
 
+template <typename T>
+struct LogarithmFunction {
+  FOLLY_ALWAYS_INLINE bool call(double& result, double a, double b) {
```
Comment (rui-mo) [inline velox/functions/sparksql/Arithmetic.h:267]: Could we check if this implementation gives consistent result with Spark when there is special value like NaN, Inf and -Inf? Thanks.

**Review by rui-mo — COMMENTED** [review]
Thanks.

File: velox/functions/sparksql/Arithmetic.h (hunk 4 above)
Comment (zhli1142015) [reply velox/functions/sparksql/Arithmetic.h:267]: UTs are added, thanks.
image

Comment (rui-mo) [reply velox/functions/sparksql/Arithmetic.h:267]: Thanks for checking. Could we add test for log(inf, inf)?
>spark.sql("select log(cast('Inf' as double), cast('Inf' as double))").show(false)
+---------------------------------------------+
|LOG(CAST(Inf AS DOUBLE), CAST(Inf AS DOUBLE))|
+---------------------------------------------+
|NaN                                          |
+---------------------------------------------+

File: velox/functions/sparksql/tests/ArithmeticTest.cpp [hunk 5]
```diff
@@ -582,5 +583,27 @@ TEST_F(LogNTest, log10) {
...
+    return std::isnan(res.value());
+  };
+  EXPECT_EQ(log(10, 100), 2.0);
+  EXPECT_EQ(log(0.0, 1.0), std::nullopt);
+  EXPECT_EQ(log(1.0, 0.0), std::nullopt);
+  EXPECT_EQ(log(-1.0, 1.0), std::nullopt);
+  EXPECT_EQ(log(1.0, -1.0), std::nullopt);
+  EXPECT_EQ(log(1.0, 3.0), kInf);
+  EXPECT_EQ(log(1.0, 3.0), kInf);
+  EXPECT_EQ(log(1.0, 3.0), kInf);
```
Comment (rui-mo) [inline velox/functions/sparksql/tests/ArithmeticTest.cpp:600]: Please remove duplicate tests for log(1.0, 3.0).

File: velox/functions/sparksql/Arithmetic.h (hunk 4 above)
Comment (zhli1142015) [reply velox/functions/sparksql/Arithmetic.h:267]: Added ,thanks.

File: velox/functions/sparksql/tests/ArithmeticTest.cpp (hunk 5 above)
Comment (zhli1142015) [reply velox/functions/sparksql/tests/ArithmeticTest.cpp]: Good catch, thanks.

**Review by rui-mo — APPROVED** [review]
Thanks.

**Comment by zhli1142015** [comment]
This PR is approved by @rui-mo and @PHILO-HE . @mbasmanova , @Yuhta and @pedroerp could you please help review?

**Pull request closed without merging**